### PR TITLE
[Offline Transformations] EfficientNet_b0 Pruning Transformation Enabling

### PR DESCRIPTION
**Pull request by daniil-lyakhov**
### Details:
 - Non unary activations mask propagation support
 - Fix issue branching with stop ops on one branch leads to runtime error. 
 Now all stop ops with input masks has it's own output mask, which ruin all inputs masks in case any channels were marked as prunable
 - Reduce layers support
 Reduce layer propagates dimensions which isn't affected by reduction and propagate as is other dimensions.
 - Logging expanded by checking initial masks with masks after mask propagation.
 This new logging step explicitly shows constants initialized with some mask which then loose their zero channels during mask propagation. This allow us to check all weights was pruned as expected, if all initialized zeros was eventually pruned. 
 
 ### Tickets:
 - 71396
 - 68162

### CODE REVIEW

**Comment by daniil-lyakhov** [comment]
Are any unit tests for new supported cases needed? @GlebKazantaev

**Comment by daniil-lyakhov** [comment]
TODO

- [ ]  Output warnings about ruined filters masks only when special debug environment key is specified

**Comment by GlebKazantaev** [comment]
Offline transformations were moved to src/common
Please rebase.

File: src/tests/ie_test_utils/common_test_utils/ngraph_test_utils.hpp [hunk 1]
```diff
@@ -56,8 +56,10 @@ class TransformationTestsF : public  CommonTestUtils::TestsCommon {
             ASSERT_NO_THROW(check_rt_info(function));
         }
 
-        auto res = comparator.compare(function, function_ref);
-        ASSERT_TRUE(res.valid) << res.message;
+        if (!m_disable_function_comparison) {
+            auto res = comparator.compare(function, function_ref);
+            ASSERT_TRUE(res.valid) << res.message;
+        }
```
Comment (daniil-lyakhov) [inline src/tests/ie_test_utils/common_test_utils/ngraph_test_utils.hpp:62]: @GlebKazantaev, it is workaround to make it possible to check accuracy on pruning transformation. Should be discussed

File: src/common/offline_transformations/src/pruning/propagate_masks.cpp [hunk 2]
```diff
@@ -582,20 +591,93 @@ class ngraph::pass::mask_propagation::PassThrough : public MatcherPass {
...
+                        any_input_with_masks = true;
+                    }
                 }
+            if (any_input_with_masks) {
+                // TODO: find out if any other operation except Result
+                // share output tesnors with previous operation
+                if (ngraph::is_type<opset6::Result>(m_output.get_node_shared_ptr()))
+                    setResultMask(m_output, output_mask);
+                else
+                    setMask(m_output, output_mask);
```
Comment (daniil-lyakhov) [inline src/common/offline_transformations/src/pruning/propagate_masks.cpp:678]: Have found a problem with `Result` operation: it share `m_tensor` attribute with previous operation, mask assign to `result` operation leads to previous operation mask override. Solve it by saving result mask by separate result operation key. Nobody needs to get mask from a result layer, but result mask should be saved to keep pruning masks in valid condition.

File: src/tests/functional/inference_engine/transformations/pruning_test.cpp [hunk 3]
```diff
@@ -17,6 +18,12 @@
 #include <ngraph/pass/manager.hpp>
 #include <inference_engine.hpp>
 
+
+#include "common_test_utils/ngraph_test_utils.hpp"
+
+#define VISUALIZE_TESTS_TREE false
+#define VISUALIZE_TREE_ROOT "/tmp/"
```
Comment (daniil-lyakhov) [inline src/tests/functional/inference_engine/transformations/pruning_test.cpp:25]: Have added shortcut to visualize all functions from pruning tests

**Comment by daniil-lyakhov** [comment]
Jenkins please retry a build

File: src/common/offline_transformations/include/mask_attribute.hpp [hunk 4]
```diff
@@ -203,4 +210,10 @@ Mask::Ptr getMask(const Output<Node> & output);
 
 void setMask(Output<Node> output, const Mask::Ptr & mask);
 
+void setResultMask(Output<Node> output, const Mask::Ptr & mask);
+
+Mask::Ptr getInitMask(const Output<Node> & output);
```
Comment (alexsu52) [inline src/common/offline_transformations/include/mask_attribute.hpp:215]: It looks like the initialization mask is only used for debugging. If so, create an initialization mask only in debug mode.

**Comment by daniil-lyakhov** [comment]
TODO:

- [ ] Adjust reference graph in pruning tests
- [ ] Remove special pruning debugging key

File: src/common/offline_transformations/include/mask_attribute.hpp (hunk 4 above)
Comment (daniil-lyakhov) [reply src/common/offline_transformations/include/mask_attribute.hpp:215]: @GlebKazantaev, could you please guide me which key I should use to check if ov compiled in debug mode?

File: src/common/offline_transformations/src/pruning/propagate_masks.cpp (hunk 2 above)
Comment (daniil-lyakhov) [reply src/common/offline_transformations/src/pruning/propagate_masks.cpp:678]: CC: @GlebKazantaev

File: src/common/offline_transformations/include/mask_attribute.hpp (hunk 4 above)
Comment (GlebKazantaev) [reply src/common/offline_transformations/include/mask_attribute.hpp:215]: Usually it is `ENABLE_OPENVINO_DEBUG`

Comment (GlebKazantaev) [reply src/common/offline_transformations/include/mask_attribute.hpp:215]: Also please add description to this functoins.

Comment (daniil-lyakhov) [reply src/common/offline_transformations/include/mask_attribute.hpp:223]: Thanks! Please check

**Review by alexsu52 — APPROVED** [review]

**Comment by daniil-lyakhov** [comment]
TODO:

- [ ] Save mask in result node instead of output tensor special key

**Comment by daniil-lyakhov** [comment]
Problems with face-detection-0205 related to this pr and described in ticket 75020

File: src/common/offline_transformations/src/pruning/propagate_masks.cpp [hunk 5]
```diff
@@ -582,19 +591,93 @@ class ngraph::pass::mask_propagation::PassThrough : public MatcherPass {
...
+                    cur_mask->copy_value_from_mask(output_mask_row);
+                    return true;
+                }, output_mask);
+                output_mask->add_callback([input_mask_row, reduce_dims](Mask::Ptr cur_mask) -> bool{
+                    // Propagate masks through dimension only if this dimension isn't reduced
+                    for (size_t dim = 0; dim < std::min(cur_mask->size(), input_mask_row->size()); ++dim)
+                        if (std::find(reduce_dims.begin(), reduce_dims.end(), dim) == reduce_dims.end())
+                            cur_mask->at(dim) = input_mask_row->at(dim);
+                        else if (cur_mask->at(dim) != input_mask_row->at(dim))
+                            cur_mask->initialize_dependencies();
```
Comment (daniil-lyakhov) [inline src/common/offline_transformations/src/pruning/propagate_masks.cpp:627]: This is needed in cases when masks are propagating "down" through reduce layer. For example(PruneStopReduceLayerDown test case), convolution 28043 produces non empty mask, and on step of mul op processing, this zeros goes to convolution 28030, which unsuccessfully would try to set this zeros to reduce layer. But the problem is that as reduce layer mask didn't changed thus it don't notify and update their neighbors thus zero in convolution 6902 will be kept.

image

**Review by GlebKazantaev — APPROVED** [review]

**Pull request merged**